tests/instancetype/upgrade.go

```diff
@@ -0,0 +1,529 @@
+		Expect(vm.Spec.Preference).ToNot(BeNil())
+		Expect(vm.Spec.Preference.RevisionName).ToNot(BeEmpty())
+		return vm.Spec.Preference.RevisionName
+	}
+
+	BeforeEach(func() {
+		// We create a fake instance type here just to allow for the
+		// creation of the initial VirtualMachine. This then allows the
+		// creation of a ControllerRevision later on in the test to use the
+		// now created VirtualMachine as an OwnerReference.
+		instancetype := newVirtualMachineInstancetype(nil)
+		instancetype, err := virtClient.VirtualMachineInstancetype(testsuite.GetTestNamespace(instancetype)).Create(context.Background(), instancetype, metav1.CreateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		preference := newVirtualMachinePreference()
+		preference, err = virtClient.VirtualMachinePreference(testsuite.GetTestNamespace(preference)).Create(context.Background(), preference, metav1.CreateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		vmi := libvmi.NewCirros()
+		removeResourcesAndPreferencesFromVMI(vmi)
+
+		vm = tests.NewRandomVirtualMachine(vmi, false)
+		vm.Spec.Instancetype = &virtv1.InstancetypeMatcher{
+			Name: instancetype.Name,
+			Kind: instancetypeapi.SingularResourceName,
+		}
+		vm.Spec.Preference = &virtv1.PreferenceMatcher{
+			Name: preference.Name,
+			Kind: instancetypeapi.SingularPreferenceResourceName,
+		}
+
+		vm, err = virtClient.VirtualMachine(testsuite.GetTestNamespace(instancetype)).Create(context.Background(), vm)
+		Expect(err).ToNot(HaveOccurred())
+
+		// Wait for the initial revisionNames to be populated before we start out tests
+		libinstancetype.WaitForVMInstanceTypeRevisionNames(vm.Name, virtClient)
+	})
+
+	DescribeTable("ControllerRevisionUpgrade should upgrade", func(generateControllerRevision func() (*appsv1.ControllerRevision, error), updateMatcher func(string), getVMRevisionName func() string) {
+		// Capture the original RevisionName
+		originalRevisionName := getVMRevisionName()
+
+		By("Generating the target ControllerRevision")
+		cr, err := generateControllerRevision()
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Updating the VirtualMachine to reference the generated ControllerRevision")
+		OriginalTestRevisionName := cr.Name
+		updateMatcher(OriginalTestRevisionName)
+
+		By("Waiting for the ControllerRevision referenced by the VirtualMachine to be upgraded to the latest version")
+		var revisionName string
+		Eventually(func(g Gomega) {
+			By("Waiting for the RevisionName to be updated")
+			revisionName = getVMRevisionName()
+			g.Expect(revisionName).ToNot(Equal(originalRevisionName))
+
+			cr, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Get(context.Background(), revisionName, metav1.GetOptions{})
+			g.Expect(err).ToNot(HaveOccurred())
+
+			By("Ensuring the referenced ControllerRevision has the latest version label")
+			g.Expect(cr.Labels).To(HaveKeyWithValue(instancetypeapi.ControllerRevisionObjectVersionLabel, instancetypeapi.LatestVersion))
```

Comment:
Also assert the correct name with updated version?

Reply:
Nope, the name is just a convenience, I don't want us to rely on it.

Reply:
ACK